File: 백준/Gold/14502. 연구소/연구소.cc
#include<bits/stdc++.h>
using namespace std;
int n, m, a[10][10], visited[10][10], ret = -1;
vector<pair<int, int>> vir, wall;
const int dy[] = {0, 1, 0, -1};
const int dx[] = {-1, 0, 1, 0};
void dfs(int y, int x){
    visited[y][x] = 1;
    for(int i = 0; i < 4; i ++){
        int ny = y + dy[i];
        int nx = x + dx[i];
        if(ny < 0 || nx < 0 || ny >= n || nx >= m) continue;
        if(a[ny][nx] == 1 || visited[ny][nx]) continue;
        dfs(ny, nx);
    }
}
void solve(){
    int cnt = 0;
    memset(visited, 0, sizeof(visited));
    for(auto v : vir) dfs(v.first, v.second);
    for(auto w : wall){
        if(a[w.first][w.second] == 0 && visited[w.first][w.second] == 0){
            cnt++;
        }
    }
    ret = max(ret, cnt);
}
int main(){
    cin >> n >> m;
    for(int i = 0; i < n; i++){
        for(int j = 0; j < m; j++){
            cin >> a[i][j];
            if(a[i][j] == 2) vir.push_back({i, j});
            if(a[i][j] == 0) wall.push_back({i, j});
        }
    }
    for(int i = 0; i <  wall.size(); i++){
        for(int j = 0; j < i; j++){
            for(int k = 0; k < j; k++){
                a[wall[i].first][wall[i].second] = 1;
                a[wall[j].first][wall[j].second] = 1;
                a[wall[k].first][wall[k].second] = 1;
                solve();
                a[wall[i].first][wall[i].second] = 0;
                a[wall[j].first][wall[j].second] = 0;
                a[wall[k].first][wall[k].second] = 0;
            }
        }
    }
    cout << ret << '\n';
}

File: 백준/Gold/14502. 연구소/연구소.py
def dfs(y, x):
    for i in range(4):
        ny = y + dy[i]
        nx = x + dx[i]
        if ny < 0 or ny >= n or nx < 0 or nx >= m or visited[ny][nx] or a[ny][nx] == 1:
            continue
        visited[ny][nx] = 1
        dfs(ny, nx)

def solve():
    # visited 배열 초기화
    for i in range(n):
        for j in range(m):
            visited[i][j] = 0
    
    # 바이러스 시작점에서 DFS 실행
    for y, x in virus_list:
        visited[y][x] = 1
        dfs(y, x)
    
    # 안전 영역 크기 계산
    cnt = 0
    for i in range(n):
        for j in range(m):
            if a[i][j] == 0 and not visited[i][j]:
                cnt += 1
    return cnt

# 방향 배열 (상, 우, 하, 좌)
dy = [-1, 0, 1, 0]
dx = [0, 1, 0, -1]

# 입력 받기
n, m = map(int, input().split())
a = [[0] * m for _ in range(n)]
visited = [[0] * m for _ in range(n)]
virus_list = []
wall_list = []

# 맵 입력 및 바이러스, 벽 위치 저장
for i in range(n):
    row = list(map(int, input().split()))
    for j in range(m):
        a[i][j] = row[j]
        if a[i][j] == 2:
            virus_list.append((i, j))
        if a[i][j] == 0:
            wall_list.append((i, j))

# 최대 안전 영역 크기 계산
ret = 0
for i in range(len(wall_list)):
    for j in range(i):
        for k in range(j):
            # 벽 3개 세우기
            a[wall_list[i][0]][wall_list[i][1]] = 1
            a[wall_list[j][0]][wall_list[j][1]] = 1
            a[wall_list[k][0]][wall_list[k][1]] = 1
            
            # 안전 영역 크기 계산
            ret = max(ret, solve())
            
            # 벽 복원
            a[wall_list[i][0]][wall_list[i][1]] = 0
            a[wall_list[j][0]][wall_list[j][1]] = 0
            a[wall_list[k][0]][wall_list[k][1]] = 0

print(ret)